src/cargo/core/shell.rs

```diff
@@ -147,27 +147,61 @@ impl MultiShell {
 }
 
 impl Shell {
-    pub fn create(out: Box<Write + Send>, config: ShellConfig) -> Shell {
+    pub fn create(out: Box<Write + Send>, config: ShellConfig) -> CargoResult<Shell> {
```

Comment:
Perhaps this could take `FnMut() -> Box<Write + Send>` and just call that multiple times if necessary? The actual handles are just stdout/stderr so they can be manufactured multiple tiems.